### PR TITLE
Implement the skeleton of the OptimalControlElement class and SE3Task

### PR DESCRIPTION
**Pull request by GiulioRomualdi**
This PR implements the first skeleton for the TSID library. It is the first PR that will allow achieving #166. In detail, I defined an `OptimalControlElement` that is an abstract class that described a generic task that can be achieved by TSID. I also developed `SE3Task` whose purpose is to implement `J dν/dt + dJ/dt ν = dv/dt*`.

[`lie-group-controllers`](https://github.com/dic-iit/lie-group-controllers) is now added as a dependency of `blf`. Once this PR has been merged `lie-group-controllers` should be added to the superbuild (cc @traversaro )

Since I would like to keep the PR small, I will implement the other tasks into other PR.

### CODE REVIEW

**Comment by GiulioRomualdi** [comment]
image
lovelly!

**Comment by GiulioRomualdi** [comment]
On macOS is failing because of 
```
==> Installing dependencies for pybind11: python@3.9
==> Installing pybind11 dependency: python@3.9
==> Pouring python@3.9-3.9.1_5.catalina.bottle.tar.gz
Error: The `brew link` step did not complete successfully
The formula built, but is not symlinked into /usr/local
Could not symlink bin/2to3
Target /usr/local/bin/2to3
already exists. You may want to remove it:
  rm '/usr/local/bin/2to3'
```

**Comment by traversaro** [comment]
> 
> 
> On macOS is failing because of
> 
> ```
> ==> Installing dependencies for pybind11: python@3.9
> ==> Installing pybind11 dependency: python@3.9
> ==> Pouring python@3.9-3.9.1_5.catalina.bottle.tar.gz
> Error: The `brew link` step did not complete successfully
> The formula built, but is not symlinked into /usr/local
> Could not symlink bin/2to3
> Target /usr/local/bin/2to3
> already exists. You may want to remove it:
>   rm '/usr/local/bin/2to3'
> ```

See:
* https://github.com/actions/virtual-environments/issues/2391
* https://github.com/actions/virtual-environments/issues/2322

In short, the combination of homebrew and GitHub Actions is extremely unstable, and incidents like that are quite frequent. Once strategy may be probably to migrate to conda-forge provided dependencies at least for macOS.

**Comment by traversaro** [comment]
> [`lie-group-controllers`](https://github.com/dic-iit/lie-group-controllers) is now added as a dependency of `blf`. Once this PR has been merged `lie-group-controllers` should be added to the superbuild (cc @traversaro )

Done https://github.com/robotology/robotology-superbuild/pull/574 .

**Comment by DanielePucci** [comment]
CC @diegoferigo

**Comment by prashanthr05** [comment]
@GiulioRomualdi I will take a look at this PR soon. But anyhow,  for starters, a naming clarification.
If I am not wrong `SE3Task` is technically not a $SE(3)$ space defined task, instead is a $SO(3) \times \mathbb{R}^3$ task? 

Theoretically, there are differences  between the two due to the different definitions of exponential maps and logarithm maps. 

Also in the implementation, at a higher level you set the pose, but at the lower level you instead split it into a rotation and a position object to perform the interpolation. 

So in that case, do you think we should rename the class?

File: src/TSID/src/SE3Task.cpp [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+                                     const manif::SE3d::Tangent& mixedVelocity,
+                                     const manif::SE3d::Tangent& mixedAcceleration)
+{
+
+    bool ok = true;
+    ok = ok && m_R3Controller.setDesiredState({I_H_F.translation(), mixedVelocity.v()});
+    ok = ok && m_R3Controller.setFeedForward(mixedAcceleration.v());
+
+    ok = ok && m_SO3Controller.setDesiredState({I_H_F.quat(), mixedVelocity.w()});
+    ok = ok && m_SO3Controller.setFeedForward(mixedAcceleration.w());
```
Comment (prashanthr05) [inline src/TSID/src/SE3Task.cpp:142]: @GiulioRomualdi, Just to notify you of an upcoming change of these getter method names in https://github.com/artivis/manif/pull/190.

File: src/TSID/include/BipedalLocomotion/TSID/OptimalControlElement.h [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+ * \f$A\f$ and a vector \f$b\f$. This class describes both a linear equality constraint and a linear
+ * inequality constraint. In case of equality constraint \f$ A \f$ and \f$ b \f$ represents: \f$ Ax
+ * = b\f$ In case of inequality constraint \f$ A \f$ and \f$ b \f$ represents: \f$ Ax \le b \f$
+ * @note Please inherit this class if you want to build your own optimal control problem.
+ */
+class OptimalControlElement
+{
+protected:
+    Eigen::MatrixXd m_A; /**< Element Matrix */
+    Eigen::VectorXd m_b; /**< Element Vector */
```
Comment (S-Dafarra) [inline src/TSID/include/BipedalLocomotion/TSID/OptimalControlElement.h:36]: I would suggest to name this in a way that is more specific to TSID, like ``TaskElement``, hiding how that TSID problem is solved (either via optimal control, pseudoinverse, or other methods). ``OptimalControl`` is, in my opinion, generic and a bit misleading, considering that you are considering only linear elements.

Comment (GiulioRomualdi) [reply src/TSID/include/BipedalLocomotion/TSID/OptimalControlElement.h]: here: https://github.com/dic-iit/bipedal-locomotion-framework/pull/167/commits/f7cd662288b4c945ea3dbc8be285968027117feb I renamed the class as `Task`

**Review by S-Dafarra — APPROVED** [review]

**Comment by GiulioRomualdi** [comment]
> @GiulioRomualdi I will take a look at this PR soon. But anyhow, for starters, a naming clarification.
> If I am not wrong `SE3Task` is technically not a SE(3) space defined task, instead is a SO(3)×R3 task?

Yes exactly

> Theoretically, there are differences between the two due to the different definitions of exponential maps and logarithm maps.
> 
> Also in the implementation, at a higher level you set the pose, but at the lower level you instead split it into a rotation and a position object to perform the control. (This is misleading).
> 
> So in that case, do you think we should rename the class?

I tried to follow what was done in [`TSID`](https://github.com/stack-of-tasks/tsid/blob/master/src/tasks/task-se3-equality.cpp). However, I agree with you that the name could be misleading. Indeed here the MIXED representation is used to define the twist and as you know in Lie theory only left/right trivializations are well defined. 

Do you have other names in mind?

**Comment by prashanthr05** [comment]
> Do you have other names in mind?

A wordy choice would be in the direction of `SO3xR3TaskMixed`. where the prefix for Task will specify the task space and the suffix will specify the velocity representation ?

**Comment by traversaro** [comment]
> > Do you have other names in mind?
> 
> A wordy choice would be in the direction of `SO3xR3TaskMixed`. where the prefix for Task will specify the task space and the suffix will specify the velocity representation ?

I think that even if not mathematically super-precise, `SE3Task` may be a good compromise.

**Comment by GiulioRomualdi** [comment]
> > Do you have other names in mind?
> 
> A wordy choice would be in the direction of `SO3xR3TaskMixed`. where the prefix for Task will specify the task space and the suffix will specify the velocity representation ?
> 
> Or suppress the suffix, so that different choices are handled dynamically within the class (in the future), as done in `iDynTree` for some computations.

In https://github.com/dic-iit/bipedal-locomotion-framework/pull/167/commits/7a776b256581da2ee34de5539b10d1f54875c507 I added a check to make sure that the `KinDynComputations` object handles quantities in mixed representation. About changing the name of the class I would like to keep it as shorter as possible. If you agree I can improve the documentation by explaining the "theoretical" issue you spot.

**Review by prashanthr05 — COMMENTED** [review]
minor comment.

**Comment by prashanthr05** [comment]
> @prashanthr05
> 
> > > Do you have other names in mind?
> > 
> > 
> > A wordy choice would be in the direction of `SO3xR3TaskMixed`. where the prefix for Task will specify the task space and the suffix will specify the velocity representation ?
> > Or suppress the suffix, so that different choices are handled dynamically within the class (in the future), as done in `iDynTree` for some computations.
> 
> In [7a776b2](https://github.com/dic-iit/bipedal-locomotion-framework/commit/7a776b256581da2ee34de5539b10d1f54875c507) I added a check to make sure that the `KinDynComputations` object handles quantities in mixed representation. About changing the name of the class I would like to keep it as shorter as possible. If you agree I can improve the documentation by explaining the "theoretical" issue you spot.

Yes. I think it's ok.

File: src/TSID/src/SE3Task.cpp [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+    if (!ptr->getParameter("robot_acceleration_variable_name", robotAccelerationVariableName)
+        || !variablesHandler.getVariable(robotAccelerationVariableName,
+                                         m_robotAccelerationVariable))
+    {
+        std::cerr << errorPrefix << descriptionPrefix << frameName
+                  << " - Error while retrieving the robot acceleration variable." << std::endl;
+        return false;
+    }
+
+    if (m_robotAccelerationVariable.size != m_kinDyn->getNrOfDegreesOfFreedom() + 6)
```
[the remark this answers is not in the document]
Comment (GiulioRomualdi) [reply src/TSID/src/SE3Task.cpp]: Done https://github.com/dic-iit/bipedal-locomotion-framework/pull/167/commits/bfa5eaa1f2d7611911eda7681d6e5960c241078d

**Comment by GiulioRomualdi** [comment]
@prashanthr05

> > > > Do you have other names in mind?
> > > 
> > > 
> > > A wordy choice would be in the direction of `SO3xR3TaskMixed`. where the prefix for Task will specify the task space and the suffix will specify the velocity representation ?
> > > Or suppress the suffix, so that different choices are handled dynamically within the class (in the future), as done in `iDynTree` for some computations.
> > 
> > 
> > In [7a776b2](https://github.com/dic-iit/bipedal-locomotion-framework/commit/7a776b256581da2ee34de5539b10d1f54875c507) I added a check to make sure that the `KinDynComputations` object handles quantities in mixed representation. About changing the name of the class I would like to keep it as shorter as possible. If you agree I can improve the documentation by explaining the "theoretical" issue you spot.
> 
> Yes. I think it's ok.

https://github.com/dic-iit/bipedal-locomotion-framework/pull/167/commits/72041a72be4e4fb4ff0efb15d0cb0b2b3b646625 let me know if it's fine with you.

**Comment by prashanthr05** [comment]
> @prashanthr05
> 
> > > > > Do you have other names in mind?
> > > > 
> > > > 
> > > > A wordy choice would be in the direction of `SO3xR3TaskMixed`. where the prefix for Task will specify the task space and the suffix will specify the velocity representation ?
> > > > Or suppress the suffix, so that different choices are handled dynamically within the class (in the future), as done in `iDynTree` for some computations.
> > > 
> > > 
> > > In [7a776b2](https://github.com/dic-iit/bipedal-locomotion-framework/commit/7a776b256581da2ee34de5539b10d1f54875c507) I added a check to make sure that the `KinDynComputations` object handles quantities in mixed representation. About changing the name of the class I would like to keep it as shorter as possible. If you agree I can improve the documentation by explaining the "theoretical" issue you spot.
> > 
> > 
> > Yes. I think it's ok.
> 
> [72041a7](https://github.com/dic-iit/bipedal-locomotion-framework/commit/72041a72be4e4fb4ff0efb15d0cb0b2b3b646625) let me know if it's fine with you.

Great, thanks!!

**Review by prashanthr05 — APPROVED** [review]

**Pull request merged**